### ISSUE TITLE
Image stack to dataset process can exhaust ram for large datasets

### ISSUE THREAD

**Issue by pgunn** [scalableminds/webknossos-libs#1146]
I'm following these instructions to make a wkw of our segmentation data stack:
https://docs.webknossos.org/webknossos-py/examples/image_stack_to_dataset.html

This is a set of about 8500 uint64 tiffs of dimensionality 9215x12799

This unfortunately exhausts the ram of my (fairly beefy - 256G) workstation; I'm guessing that the dataset library is building an array in memory before writing rather than saving data bit-by-bit.

Is this the best way to prepare this kind of segmentation dataset for import to your web platform, or is a zipfile of tiffs better?

**Comment by philippotto** [comment]
Hi @pgunn,

could you try passing `batch_size=32` to `add_layer_from_images`? Also ensure that you **don't** pass `compress=True`. WKW is by default written in shards of size 1024^3, where each shard is chunked into 32^3 blocks. By default, sections are buffered until the current shard is full. For your dataset, this would require too much RAM. By setting the batch_size to 32, the RAM requirement should be lower by a factor of 32.

Please let me know if this helps.

Thank you!

**Comment by pgunn** [comment]
I tried a chunk size of 32 but it still came pretty close to eating all my RAM and I killed it before other bad things could happen.

I wanted to try a batch size of 16, but I got a message that batch_size 16 must be divisible by z chkink-size 32

(I was puzzled by the "batch_size 32 reducing the RAM requirement by a factor of 32 ; is a higher batch size actually what I should want? Seems counterintuitive)

**Comment by pgunn** [comment]
The library also may have a race condition of some sort WRT parallelism; a lot of the time I get a message that the export directory isn't empty even though I've verified that it is before starting.

```
scclin022(1):/mnt/ceph/data/neuro/wasp_em/segmentation_evaluation/eval1$ rm -rf export/ ; mkdir export
scclin022:/mnt/ceph/data/neuro/wasp_em/segmentation_evaluation/eval1$ ~/src/waspem-tooling/data_managers/tiffstack_to_wkw stack export --batch_size 64
Traceback (most recent call last):
  File "/mnt/home/pgunn/src/waspem-tooling/data_managers/tiffstack_to_wkw", line 34, in <module>
    main()
  File "/mnt/home/pgunn/src/waspem-tooling/data_managers/tiffstack_to_wkw", line 18, in main
    dataset = webknossos.Dataset(dataset_path=cfg.dest, voxel_size=(8,8,8), name=cfg.dset_name, exist_ok=False)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/pgunn/miniconda3/envs/webknossos/lib/python3.11/site-packages/webknossos/dataset/dataset.py", line 288, in __init__
    raise RuntimeError(
RuntimeError: Creation of Dataset at export failed, because a non-empty folder already exists at this path.
scclin022(1):/mnt/ceph/data/neuro/wasp_em/segmentation_evaluation/eval1$ 
```

**Comment by philippotto** [comment]
> I tried a chunk size of 32 but it still came pretty close to eating all my RAM and I killed it before other bad things could happen.

Could you please try the following experiment: Create a folder that only contains 32 tiffs and start your script with that. Does it exceed the RAM, too?

> (I was puzzled by the "batch_size 32 reducing the RAM requirement by a factor of 32 ; is a higher batch size actually what I should want? Seems counterintuitive)

By default, the batch_size equals the shard_size (which is by default 1024). Therefore, 32 is a smaller value :)

> The library also may have a race condition of some sort WRT parallelism; a lot of the time I get a message that the export directory isn't empty even though I've verified that it is before starting.

This is weird. Are you on a network storage? As a workaround, you could provide a fresh dataset name for each invocation. By the way, you don't need to create the folder before instantiating `webknossos.Dataset`.

**Comment by pgunn** [comment]
Ah, I didn't spot offhand that 32*32 is 1024.

I am on network storage (ceph); maybe the consistency model doesn't play nice with this code; I can point it at local storage to avoid these issues and copy it back later.

With a directory having only 32 images, it ran to completion (although the progress bar remained at 0% until it went to 100% at the very end):
```
scclin022:/mnt/ceph/data/neuro/wasp_em/segmentation_evaluation/eval1$ ~/src/waspem-tooling/data_managers/tiffstack_to_wkw stack32 export32
Creating layer seg from images ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:06:06 | 0:00:00
scclin022:/mnt/ceph/data/neuro/wasp_em/segmentation_evaluation/eval1$ 
```

I'm wondering if the batch_size isn't being respected properly with add_layer_from_images; if there's some other flags I need to handle large data I'm happy to experiment.

(the end goal of all this is to let us evaluate proofreading on webknossos for our data; I hope making a wkw and uploading that is the right route forward rather than just uploading a zip of our tiffs; I'm open to alternative approaches)

**Comment by philippotto** [comment]
I think the problem is that the conversion is parallelized with n=cpu_count by default and it's a bit cumbersome to tweak via code (the CLI interface `webknossos convert` provides a `--jobs` parameter though).
Please try to adapt your script with this snippet:

```
from argparse import Namespace

def main():
    ...
    executor_args = Namespace(
        jobs=1,
        distribution_strategy="multiprocessing",
    )
    with wk.utils.get_executor_for_args(None) as executor:
        dataset.add_layer_from_images(..., executor=executor)
```

A value higher than 1 should probably work too depending on the actual RAM consumption, but let's try it with 1 first.

> (although the progress bar remained at 0% until it went to 100% at the very end):

Yes, the progress is only computed chunk-wise. In your case there was only one chunk which explains the choppy progress.

> (the end goal of all this is to let us evaluate proofreading on webknossos for our data; I hope making a wkw and uploading that is the right route forward rather than just uploading a zip of our tiffs; I'm open to alternative approaches)

Good to hear! Don't hesitate to reach out if you have any questions, problems or feature requests. Proofreading is a feature for which we are currently pushing various improvements.

**Comment by pgunn** [comment]
Looking at the changes suggested, it's not clear where executor_args is read; am I missing something? I think adding the code as-is won't have that jobs=1 do anything?

**Comment by pgunn** [comment]
Do I pass executor_args in as an argument to get_executor_for_args() maybe? Instead of None?

**Comment by philippotto** [comment]
Whoops, my bad!

> Do I pass executor_args in as an argument to get_executor_for_args() maybe? Instead of None?

Yes, exactly.

**Comment by pgunn** [comment]
So far memory usage for the script is holding steady at about 34G (Probably best to leave it at this level because I have a few other RAM-expensive things I run); I'm guessing this'll be going for awhile.

I'll let you know if I run into issues or if it runs to completion, and I'll close the ticket if it does the latter. Thanks.

I imagine afterwards I can have a separate script that just initialises webknossos.Dataset against the directory and then does a push to the Webknossos servers?

**Comment by pgunn** [comment]
I had to restart targeting our ceph (shared filesystem) because I ran out of local storage; currently 2h23m in at 9%. Fingers crossed this will finish sometime tomorrow.

**Comment by philippotto** [comment]
:crossed_fingers: 
Once the conversion has finished, you should call `dataset.compress()` (before uploading). Since you are dealing with a segmentation, this should reduce the file size greatly.

**Comment by pgunn** [comment]
It finished producing the output directory without problem.

Unfortunately, calling dataset.compress() eats all my RAM and then the system falls over, even if I run it with the same executor.

**Comment by pgunn** [comment]
I initially was having trouble calling compress() because it ate all my ram and the system fell over, but then I noticed that it doesn't work with the execution context the same way add_layer_from_images() does; it instead takes the executor as an argument.

When I do this, it seems to be behaving reasonably; hopefully the compression will go to completion. Hopefully this will go smoothly. I'm guessing when I finally upload this thing I'll need to be careful to do so in a `jobs=1` way.

**Comment by philippotto** [comment]
I'm glad you got it to work!

> but then I noticed that it doesn't work with the execution context the same way add_layer_from_images() does; it instead takes the executor as an argument.

I'm not sure what you mean. `add_layer_from_images` and `compress` both take an argument named `executor`. So, all of these should work:

```
dataset.add_layer_from_images(..., executor=executor)
dataset.compress(executor)
dataset.compress(executor=executor)
```

Or do you mean something else?

-----------

> I'm guessing when I finally upload this thing I'll need to be careful to do so in a jobs=1 way.

The uploader works on chunk sizes of 100 MiB. Also, the default is not the core count (because the operation is not cpu-bound), but instead simply `5`. So, there shouldn't be a need to set it to 1. The default is probably fine.

**Comment by pgunn** [comment]
Ah, nevermind on that context thing; I must have gotten things mixed up.

Compression has been going all night; hopefully it will wrap up today.
I wonder how many other people are trying to work with datasets this big on the platform.

If it's be helpful, maybe these sample scripts would be good to toss up into an example repo for anyone else who might be trying to do this kind of thing, even knowing they might need/want to modify them.

**Comment by pgunn** [comment]
We can close this ticket; the ingest-to-file and compression worked fine. If I have issues with ingest I'll file another issue.

**Comment by pgunn** [comment]
@philippotto 
(the upload finished and the segmentation is viewable in webknossos; I now have tons of questions to get to where I'm ready to evaluate the platform four our needs, but they're outside the scope of this ticket - are there forums for this kind of thing?)

**Comment by hotzenklotz** [comment]
> are there forums for this kind of thing

https://forum.image.sc/tag/webknossos

I can also recommend the tutorials in the WK docs: https://docs.webknossos.org/webknossos/ and our YouTube channel for short "Today I learned" videos: https://www.youtube.com/@webknossos
